### PR TITLE
Add UI for warnings that occurred during loading

### PR DESCRIPTION
**Pull request by ichttt**
Currently, we only print broken files in the log, but cpw said he may want some gui to show this, so I came up with something.
There is an API for adding custom warning messages, so we can add other things to the UI quite easily as well (something like broken resources in dev e.g.). The warning screen can be disabled in the config, and allows you to continue to the main menu.
Some screenshots:
warning
warn and error
error

### CODE REVIEW

File: src/fmllauncher/java/net/minecraftforge/fml/loading/moddiscovery/ModDiscoverer.java [hunk 1]
```diff
@@ -86,8 +86,4 @@ public BackgroundScanHandler discoverMods() {
         loadingModList.setBrokenFiles(brokenFiles);
         return backgroundScanHandler;
     }
-
-    public void addExplodedTarget(final Path compiledClasses, final Path forgemodstoml) {
```
Comment (LexManos) [inline src/fmllauncher/java/net/minecraftforge/fml/loading/moddiscovery/ModDiscoverer.java]: ?

File: src/main/java/net/minecraftforge/fml/ModLoader.java [hunk 2]
```diff
@@ -186,4 +190,12 @@ public void finishMods()
         GameData.freezeData();
     }
 
+    public List<ModLoadingWarning> getWarnings() {
```
Comment (LexManos) [inline src/main/java/net/minecraftforge/fml/ModLoader.java]: style

File: src/main/java/net/minecraftforge/fml/ModLoadingWarning.java [hunk 3]
```diff
@@ -0,0 +1,60 @@
...
+
+    public ModLoadingWarning(final IModInfo modInfo, final ModLoadingStage warningStage, final String i18nMessage, Object... context) {
+        this.modInfo = modInfo;
+        this.warningStage = warningStage;
+        this.i18nMessage = i18nMessage;
+        this.context = Arrays.asList(context);
+    }
+
+    public String formatToString() {
+        return ForgeI18n.parseMessage(i18nMessage, Streams.concat(Stream.of(modInfo, warningStage), context.stream()).toArray());
```
Comment (LexManos) [inline src/main/java/net/minecraftforge/fml/ModLoadingWarning.java:58]: That seems a odd way to do it, why not just build the list in the constructor?

**Comment by LexManos** [comment]
I thought we were talking about making these warnings always on in dev. But opt-in in production? Is there anything in there?

File: src/fmllauncher/java/net/minecraftforge/fml/loading/moddiscovery/ModDiscoverer.java (hunk 1 above)
Comment (ichttt) [reply src/fmllauncher/java/net/minecraftforge/fml/loading/moddiscovery/ModDiscoverer.java]: Whoops, didn't mean to put this in this PR. It's just an unused method

File: src/main/java/net/minecraftforge/fml/ModLoadingWarning.java (hunk 3 above)
Comment (ichttt) [reply src/main/java/net/minecraftforge/fml/ModLoadingWarning.java:58]: This is more or less like the error screen does it. I think at construction langs might not be loaded yet

**Comment by ichttt** [comment]
I would argue that files that contain no mods matter to the end user. If they put e.g. an 1.12 mod or optifine in therr, currently there will only be a small log, which many users might not need. We may still add some only in-dev warnings as an addition. Currently this only shows broken files.

**Comment by LexManos** [comment]
true, thought this also showed broken resource packs/other things we discussed before.

**Comment by ichttt** [comment]
There isn't any other warning screen in 1.12 or 1.13. There is PR #4947, but it's only for 1.12 and not merged yet. And if we want dev-only warnings, we can add a check if we are in a dev enivorment before calling ModLoader#addWarning. But I think we should start with this warning and add others later on.

**Pull request merged**